python/ray/data/datasource/parquet_datasource.py

```diff
@@ -254,6 +275,47 @@ def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
 
         return read_tasks
 
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the Parquet files encoding ratio.
+
+        To avoid OOMs, it is safer to return an over-estimate than an underestimate.
+        """
+        # Sample a few rows from Parquet files to estimate the encoding ratio.
+        # Launch tasks to sample multiple files remotely in parallel.
+        # Evenly distributed to sample N rows in i-th row group in i-th file.
+        # TODO(ekl/cheng) take into account column pruning.
+        num_files = len(self._pq_ds.pieces)
+        num_samples = int(num_files * PARQUET_ENCODING_RATIO_ESTIMATE_SAMPLING_RATIO)
+        min_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MIN_NUM_SAMPLES, num_files
+        )
+        max_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MAX_NUM_SAMPLES, num_files
+        )
+        num_samples = max(min(num_samples, max_num_samples), min_num_samples)
+
+        # Evenly distributed to choose which file to sample, to avoid biased prediction
+        # if data is skewed.
+        file_samples = [
+            self._pq_ds.pieces[idx]
+            for idx in np.linspace(0, num_files - 1, num_samples).astype(int).tolist()
+        ]
+
+        ratio = 0
+        if len(file_samples) > 1:
```

Comment:
Won't it always be >1 since we have MIN_NUM_SAMPLES = 4? Can simplify and always use remote tasks in this case.

Reply:
@ericl - it actually can be smaller than `MIN_NUM_SAMPLES`, when total number of input. files are less than `MIN_NUM_SAMPLES` (per line 289-295). I am mostly worried about adding an extra hop (i.e. task launches another task) for CUJ when user just tries example in wiki to read single parquet file (`ray.data.read_parquet(<file-name>)`). WDYT?

Reply:
Running a single task should be pretty lightweight--- we already run a much larger number of tasks for other operations, so I think this is fine to simplify the code paths.

Reply:
@ericl - cool, will update.